Add Codesgiving tip of the week blog post (#27)
CU-DBMI/set-website: _posts/2023-11-15-Codesgiving-Part-1.md

---
title: "Tip of the Week: Codesgiving - Open-source Contribution Walkthrough"
author: dave-bunten
tags:
  - tip-of-the-week
  - codesgiving
---

# Tip of the Week: Codesgiving - Open-source Contribution Walkthrough

{% include tip-of-the-week-intro.html %}

## Introduction

<!-- excerpt start -->
[Thanksgiving](https://en.wikipedia.org/wiki/Thanksgiving) is a holiday practiced in many countries which focuses on gratitude for good harvests of the preceding year.

> In the United States, we celebrate Thanksgiving on November 23rd each year often by eating meals we create together with others.

Reviewer: I think it's on the fourth Thursday each year so not always the 23rd?

Author: Thanks, oops, going to correct this!

This post channels the spirit of Thanksgiving by _giving_ our thanks through _code_ as a ___"Codesgiving"___,  acknowledging and creating better software together.
<!-- excerpt end -->

## Giving Thanks to Open-source Harvests

Part of building software involves the use of code which others have built, maintained, and distributed for a wider audience.
Using other people's work often comes in the form of [open-source](https://en.wikipedia.org/wiki/Open_source) "harvesting" as we find solutions to software challenges we face.
Examples might include installing and depending upon Python packages from [PyPI](https://en.wikipedia.org/wiki/Python_Package_Index) or R packages from [CRAN](https://en.wikipedia.org/wiki/R_package#Comprehensive_R_Archive_Network_(CRAN)) within your software projects.

> "Real generosity toward the future lies in giving all to the present."
> \- Albert Camus

These open-source projects have internal costs which are sometimes invisible to those who consume them.

> Every software project has an implied level of [software gardening](https://bssw.io/blog_posts/long-term-software-gardening-strategies-for-cultivating-scientific-development-ecosystems) time costs involved to impede decay,  practice continuous improvements and evolve the work.

Author: Great catch, thanks!

    verify_check2 --> |Yes: issue\nresolved|resolved

    style reported fill:#FEF3C7,stroke:#D97706;
    style resolved fill:#BBF7D0,stroke:#16A34A;
```

_The troubleshooting process as a workflow involving looped checks for verifying an issue and validating the solution fixes an issue._

It can feel overwhelming to find a way to contribute to open-source.
Similar to other software methodology, modularizing your approach can help you progress without being overwhelmed.
Using a troubleshooting approach like the above can help you break down big challenges into bite-sized chunks.
Consider each step as a "module" or "section" which needs to be addressed sequentially.

### Embrace a Learning Mindset

Open-source contributions almost always entail learning of some kind.
Many contributions happen solely in the form of code and text communications which are easily misinterpreted.
__Assume positive intent__ and accept input from others while upholding your own ideas to share successful contributions together.
Prepare yourself by intentionally opening your mind to input from others, even if you're sure you're absolutely "right".

### Setting Software Scheduling Expectations

> 

Author: Thanks, will apply this change in a further commit!

<script src="https://cdn.jsdelivr.net/npm/vega@5.22.1"></script>
<script src="https://cdn.jsdelivr.net/npm/vega-lite@5.6.0"></script>
<script src="https://cdn.jsdelivr.net/npm/vega-embed@6.21.0"></script>
<div id="vis"></div>

<script>
  var spec = {
    "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
    "description": "Suggested Open-source Time Allocations",

>     "width": 300,

Reviewer 2: I've noticed that the legend text gets ellipsized at this width. Perhaps the entire figure or just the legend could be made wider?

It's not a blocker since you can still hover over the corresponding slice to see a tooltip with the full text (which, I might add, is a cool feature), but I thought I'd mention it anyway.

Reviewer: Related, might want to add this for smaller screen widths:

```html
<style>
canvas {
  max-width: 100%;
  height: unset;
}
</style>
```

Or if vega allows `width` to be "auto" (full width), I think that'd be fine. Or a max-width of 300px.

Author: Thanks - great points and suggestions! I'll see about improving this.

    "encoding": {
      "theta": {"field": "percentage", "type": "quantitative", "title": "Percentage"},
      "color": {"field": "task", "type": "nominal", "title": "Task"}
    }
  };

  const embed_opt = {"mode": "vega-lite"};
  const el = document.getElementById('vis');
  const view = vegaEmbed("#vis", spec, embed_opt);
</script>
_Suggested ratio of time spent by type of work for an open-source contribution._
{:.center}

The following modified rule of thumb from [_The Mythical Man Month_](https://www.oreilly.com/library/view/mythical-man-month-the/0201835959/) can assist with how you structure your time for an open-source contribution.
Notice the emphasis on planning and testing and keep these in mind as you progress (the actual programming time can be small if adequate time has been spent on planning).
Notably, the original time fractions are modified here with the final quarter of the time spent suggested as code review, revisions, and post-actions.
Planning for the time expense of the added code review and related elements assists with keeping a learning mindset throughout the process (instead of feeling like the review is a "tack-on" or "optional / supplementary").

> A good motto to keep in mind throughout this process is [_Festina lente_](https://en.wikipedia.org/wiki/Festina_lente), or __"Make haste, slowly."__ (take care to move thoughtfully and as slow as necessary to do thing correctly the first time).

Reviewer 2: ```suggestion
A good motto to keep in mind throughout this process is [_Festina lente_](https://en.wikipedia.org/wiki/Festina_lente), or __"Make haste, slowly."__ (take care to move thoughtfully and as slowly as necessary to do things correctly the first time).
```

Author: Thanks, reads better this way!

> 1. 1/3 planning (~33%)

Reviewer: Perhaps move this to right below the chart, since it's directly related? Good on you for putting it in text, as these data visualizations are seldom accessible.

Author: Great idea, thanks! I'll move this as you mentioned.

2. 1/6 coding (~16%)
3. 1/4 component and system testing (25%)
4. 1/4 code review, revisions, and post-actions (25%)

## Planning an Open-source Contribution

### Has the Need Already Been Reported?

Be sure to check whether the bug or feature has already been reported somewhere!
In a way, this is a practice of ["Don't repeat yourself" (DRY)](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) where we attempt to avoid repeating the same block of code (in this case, the "code" can be understood as natural language).

> For example, you can look on GitHub Issues with a search query matching the rough idea of what you're thinking about.

Reviewer: Very minor point, but maybe mention GitHub Discussions too, because I feel like people forget about this usually.

Author: Such a good point, thanks! I'll find a way to work this in. I wonder if there's a way to mention the new search bar functionality to this effect (unsure if it highlights discussions alongside issues).

If it has been reported already, take a look to see if someone has made a code contribution related to the work already.

> If there aren't yet any code contributions and it doesn't look like anyone is working on one, consider volunteering to take a further look into the solution and be sure to acknowledge the existing discussions.

Reviewer: Another very minor point, but I find that it helps to acknowledge the design difficulties and constraints of the author/maintainer, e.g. "i can see how creating a flexible and simple API would be hard for this, but maybe you could do something like this".

Author: Love this, showing sympathy or empathy can go a long way. Thanks - I'll find a way to add this in.

### Reporting the Need for Change

```markdown
# Using `function_x` with `library_y` causes `exception_z`

## Summary

As a `library_y` research software developer I want to use `function_x` for my data so that I can share data for research outcomes.

## Reproducing the error

This error may be seen using Python v3.x on all major OS's using the following code snippet:
...

```

_An example of a user story issue report with imagined code example._

Open-source needs are often best reported through written stories captured within a bug or feature tracking system (such as [GitHub Issues](https://github.com/features/issues)) which if possible also include example code or logs.
One template for reporting issues is through a "user story".
A user story typically comes in the form: `As a < type of user >, I want < some goal > so that < some reason >.` ([Mountain Goat Software: User Stories](https://www.mountaingoatsoftware.com/agile/user-stories)).
Alongside the story, it can help to add in a snippet of code which exemplifies a problem, new functionality, or a potential adjacent / similar solution.

> As a general principle, be as __be as specific as you can without going overboard__.

Reviewer 2: ```suggestion
As a general principle, __be as specific as you can without going overboard__.
```

Author: Thanks!

Include things like programming language version, operating system, and other system dependencies that might be related.

> Once you have a good written description of the need be sure to submit it where it can be seen by the relevant development community.

Reviewer 2: ```suggestion
Once you have a good written description of the need, be sure to submit it where it can be seen by the relevant development community.
```

Author: Great catch!

## Developing and Testing an Open-source Contribution

### Creating a Development Workspace

Once ready to develop a solution for the reported need in the open-source project you'll need a place to version your updates.
This work generally takes place through version control on focused branches which are named in a way that relates to the focus.
When working on GitHub, this work also commonly takes place on [forked repository copies](https://docs.github.com/en/get-started/quickstart/fork-a-repo).
Using these methods helps isolate your changes from other work that takes place within the project.
It also can help you track your progress alongside related changes that might take place before you're able to seek review or code merges.

#### Bug or Feature Verification with Test-driven Development

> 1. Add or modify a test which checks for a bug fix or feature addition
> 2. Run all tests (expecting the newly added test content to fail)
> 3. Write a simple version of code which allows the tests to succeed
> 4. Verify that all tests now pass
> 5. Return to step 3, refactoring the code as needed

> 

Reviewer 2: Are these steps supposed to be below the line further down that says "Test-driven development entails the following pattern:"?

Author: Whoops, yes, thanks! I tried to make these steps a sort of "graphic" ahead of the text, but they lost the improved flow you mention. I'll work this in somehow with an additional commit for this PR.

Use [timeboxing techniques](https://cu-dbmi.github.io/set-website/2023/01/17/Timebox-Your-Software-Work.html) to  help make sure the time you spend in development is no more than necessary.
Also recall that a "simplest version" of the code can often be more quickly refactored and completed than devising a "perfect" solution the first time.
Remember, you'll also very likely have the help of a code review before the code is merged (expect to learn more and add changes during review!).

## Code Review, Revisions, and Post-actions

### Pull Requests and Code Review

When your code and new test(s) are in a good spot it's time to ask for a code review.
Be sure to follow the open-source [`CONTRIBUTING.md` document](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors) documentation related to the project; each one can vary.
When working on GitHub-based projects, you'll need to open a [pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests) on the correct branch (usually upstream `main`).

> If you used a GitHub issue to help report the issue, mention the issue in the pull request description using the `#<issue number` reference to help link the work to the reported need (this will cause the pull request to show up within the issue).

Reviewer: `#issue number`?

Author: Thanks, much better!

> Note: It might feel tempting to perfect the code.

Reviewer: I feel like this point is said several times in article. Maybe consider reviewing if any of the instances can be removed. Though it is certainly a good thing to emphasize multiple times, and keep in mind at all times during the process.

Code review takes advantage of a strength of our species: collaborative & multi-perspectival thinking.
Leverage this in your open-source experience by seeking feedback when things feel "good enough".

### Code Revisions

> "Perfection is achieved, not when there is nothing more to add, but when there is nothing left to take away."
> \- Antoine de Saint-Exupery

You may be asked to update your code based on automated code quality checks or reviewer request.
Treat these with care; embrace learning and remember that this step can take 25% of the total time for the contribution.
When working on GitHub forks or branches, you can make additional commits directly on the development branch which was used for the pull request.
If your reviewers requested changes, re-request their review once changes have been made to help let them know the code is ready for another look.

### Post-actions and Tidying Up Afterwards

Once the code has been accepted by the reviewers and through potential automated testing suite(s) the content is ready to be merged.
Oftentimes this work is completed by core maintainers of the project.

> After the code is merged, be sure to take time to clean up your workspace by deleting your development branch and syncing with the upstream repository (optional).

Reviewer: "it's usually a good idea to clean up..." rather than "(optional)"?

Author: Thanks, that sounds much better - I'll work this in!
